### ISSUE TITLE
NPM broken in mingw64_shell (module search)

### ISSUE THREAD

**Issue by radioflash** [Alexpux/MINGW-packages#1039]
When launching NPM, everything works from the default windows CMD line (with the mingw64/bin directory in PATH).

But when started from within a  mingw64_shell, npm tries to search for modules in a directory that does not exist.

Actual module path: 
```
$ echo $NODE_PATH
c:\msys64\mingw64\lib\node_modules
```

Path where npm searches for modules (should be: c:/msys64/mingw64/[...] but we get c:/mingw64/[...] instead. npm *fails* to launch if modules are not copied into this directory):
```
$ npm --help
npm@1.4.28 C:\mingw64\lib\node_modules\npm
```

npm prefix is also correct:
```
$ npm config get prefix
c:\msys64\mingw64
```

There should be no non-default npm configuration on the whole system, and NODE_PATH was set explicitly (to no effect). Any ideas on what the problem even IS, or how to fix it?

**Comment by BrisingrAerowing** [comment]
It's related to how the paths are handled. NPM seems to be taking a MSYS2 path and directly using it as a Win32 path. It is likely hardcoded somewhere in NPM, so that will need to be fixed.

**Comment by mingwandroid** [comment]
It needs to be debugged.

**Comment by radioflash** [comment]
Found it. It's a mess  :(
But there is an easy fix.

Basically, npm itself and the "executable" modules it installs are supplied as \*.cmd and as bash scripts (which feed the modules \*.js file and a path into node.exe).

The \*.cmd version (npm.cmd) works just fine (which is why everything works from a native windows command line). The shell script wrapper, on the other hand, only does the necessary path conversion if the uname output contains *CYGWIN*. This is bullshit, IMO. The "cygpath -w" path conversion should **always** be done when using node.exe AND cygpath is available.

On a sidenote: Maybe those conditions should be relaxed: cygpath -w conversion might be necessary even though the node "executable" is just "node" (not node.exe). On the other hand, that might break on weird Linux systems used for cross-compiling where cygpath is available, but should not be used. (?)

- Proper fix: modify all the wrapper scripts (affecting not only node.js, but also modules), so that the cygpath path conversion is actually done on MSYS2 as-is.
- Quick hack: Make `uname` output contain CYGWIN:
  e.g. by putting "export MSYSTEM=CYGWIN-$MSYSTEM;" into msys2/etc/bash.bashrc (not sure if MSYS2 respects ~/.bashrc by default?)

Attached file contains proposed modifications for node-shell wrapper scripts (in this case for npm, but this would apply to other modules as well). Should I go and nag at upstream maintainers to get this changed?

[npm.txt](https://github.com/Alexpux/MINGW-packages/files/129703/npm.txt)

**Comment by hannasm** [comment]
I had just started looking into this issue too. Very handy that someone has already solved it, 7 minutes earlier!

**Comment by radioflash** [comment]
Currently working on getting nodejs v5.6.0 (stable) compiled with makepkg.

This would allow us to have a nodejs-git package, and IMO it is somewhat realistic to get necessary changes integrated into upstream (>v5.6.0)-- while the only option to get this fixed for nodejs-v0.10.x (current MSYS2 version) is with an additional patch that is applied by makepkg.

What is the upgrade policy? If the current nodejs version worked with pacman (with reasonable levels of patching) would the nodejs package be updated, or are there additional hurdles? Should I work on the current nodejs LTS version (v4.3.0) instead?
